staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go

```diff
@@ -116,13 +114,43 @@ func createHandler(r rest.NamedCreater, scope *RequestScope, admit admission.Int
 
 		defaultGVK := scope.Kind
 		original := r.New()
-		trace.Step("About to convert to expected version")
-		obj, gvk, err := decoder.Decode(body, &defaultGVK, original)
+
+		validationDirective, err := fieldValidation(req, s)
 		if err != nil {
-			err = transformDecodeError(scope.Typer, err, original, gvk, body)
 			scope.err(err, w, req)
 			return
 		}
+		decodeSerializer := s.Serializer
+		if validationDirective == metav1.FieldValidationWarn || validationDirective == metav1.FieldValidationStrict {
+			decodeSerializer = s.StrictSerializer
+		}
+
+		decoder := scope.Serializer.DecoderToVersion(decodeSerializer, scope.HubGroupVersion)
+		trace.Step("About to convert to expected version")
+		obj, gvk, err := decoder.Decode(body, &defaultGVK, original)
+		if err != nil {
```

Comment:
might be more straightforward as:

```go
if err != nil {
	strictError, isStrictError := runtime.AsStrictDecodingError(err)
	switch {
	case isStrictError && validationDirective == metav1.FieldValidationWarn:
		addStrictDecodingWarnings(req.Context(), strictErr)

	default:
		err = transformDecodeError(scope.Typer, err, original, gvk, body)
		scope.err(err, w, req)
		return
	}
}
```

Reply:
when we're using FieldValidationIgnore, we didn't select to use the strict serializer, right? if so, I'd expect any error coming from the decoder to be fatal